Use diffAgainstLastAppliedFieldExclusionRules to exclude status field from diffs

praveenrewar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

If this is your first time, please read our contributor guidelines: https://github.com/vmware-tanzu/carvel-kapp/blob/develop/CONTRIBUTING.md and developer guide https://github.com/vmware-tanzu/carvel-kapp/blob/develop/docs/dev.md
-->

#### What this PR does / why we need it:
Use diffAgainstLastAppliedFieldExclusionRules to exclude status field from diffs

Acceptance criteria:
- Deploy a service or a deployment with kapp, deploy them again and there should be no diff on the status
- Try updating the status field for a service or a deployment and the diff should be seen.
- If smart diff (using last applied annotation) is not used, diff in status will be seen.

#### Which issue(s) this PR fixes:
<!--
If no issue exists for this change, please create an issue and link it here.
-->
Fixes #507 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note

```

#### Additional Notes for your reviewer:

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

sethiyash commented on pkg/kapp/config/default.go:
```diff
@@ -237,6 +230,10 @@ diffAgainstLastAppliedFieldExclusionRules:
 - path: [metadata, annotations, "deployment.kubernetes.io/revision"]
   resourceMatchers: *appsV1DeploymentWithRevAnnKey
 
+- path: [status]
```
Not super clear what it will do, the copy part is already removed above then what is the purpose of adding it here?

praveenrewar replied:
Umm, do you mean you are not clear about why I have used `diffAgainstLastAppliedFieldExclusionRules`?

sethiyash replied:
Initially was not clear but looks like this is for not to consider `status` field while diffing right?

praveenrewar replied:
Yep, that's right.

kumaritanushree replied:
What I understood is, kapp will remove `status` from lastApplied resource (resource which is currently in cluster and  used to find diff against resource we are going to apply) before calculating the diff. So that if user will add `status` in new resource then same will be shown in diff.

praveenrewar committed:
Keep original resource from cluster without rebasing

The original resource from cluster is used to determine if resource is in done state

Signed-off-by: Praveen Rewar <8457124+praveenrewar@users.noreply.github.com>

praveenrewar committed:
Add status field to diffAgainstLastAppliedFieldExclusionRules

Do not copy status from existing resource

Signed-off-by: Praveen Rewar <8457124+praveenrewar@users.noreply.github.com>

100mik commented on pkg/kapp/diff/change.go:
```diff
@@ -39,13 +40,16 @@ type ChangeImpl struct {
 var _ Change = &ChangeImpl{}
 
-func NewChange(existingRes, newRes, appliedRes ctlres.Resource) *ChangeImpl {
+func NewChange(existingRes, newRes, appliedRes, clusterOriginalRes ctlres.Resource) *ChangeImpl {
```
Why are we doing deep copies of the resources here? Is this something we should do for clusterOriginalRes as well?
Or is the intended effect of deep copying ensured?

100mik replied:
(Look at line 58)

praveenrewar replied:
Hmm, good question, I will try to see why we are doing it for others.

100mik replied:
We are not altering original res, but does not hurt to have this safeguard so I am fine with it staying

100mik commented on pkg/kapp/diff/change_precalculated.go:
```diff
@@ -51,9 +51,10 @@ func (d *ChangePrecalculated) NewOrExistingResource() ctlres.Resource {
+func (d *ChangePrecalculated) NewResource() ctlres.Resource             { return d.newRes }
+func (d *ChangePrecalculated) ExistingResource() ctlres.Resource        { return d.existingRes }
+func (d *ChangePrecalculated) AppliedResource() ctlres.Resource         { return d.appliedRes }
+func (d *ChangePrecalculated) ClusterOriginalResource() ctlres.Resource { return d.existingRes }
```
Will this ever be called? Is the existing resource for a precalculated change not rebased?

If we never run into this code path for recalculated change, should call that out here instead of returning the existing res?

praveenrewar replied:
> should call that out here instead of returning the existing res?

Do you mean we should just add a comment?

100mik replied:
I was thinking more along the lines of `panic("Unimplemented")` and a comment. So that no one down the line does not use it expecting a cluster original resource. We won't hit that if the current code does not. Or maybe `return nil` at the least because we do not actually have that information.

praveenrewar replied:
Sure, that makes sense.

praveenrewar replied:
This will be required for all the changes, it will be used in change summary.

100mik replied:
Makes sense

100mik commented on pkg/kapp/clusterapply/cluster_change.go:
```diff
@@ -146,7 +146,7 @@ func (c *ClusterChange) WaitOp() ClusterChangeWaitOp {
 		// If existing resource is not in a "done successful" state,
 		// indicate that this will be something we need to wait for
-		existingResState, _, existingErr := c.convergedResFactory.New(c.change.ExistingResource(), nil).IsDoneApplying()
+		existingResState, _, existingErr := c.convergedResFactory.New(c.change.ClusterOriginalResource(), nil).IsDoneApplying()
```
So we are using ClusterOriginalResource here so that we have access to status?
(I believe that is the whole point but just clarifying)

praveenrewar replied:
Yep, that's correct.

100mik replied:
Makes sense!

rohitagg2020 replied:
nit: change `existingResState` to `resState` or `originalResState` and `existingErr` to `err`. 

Considering that `Change` has existingRes and ClusterOriginalRes, we shouldn't mix the variable name to avoid confusion.

100mik reviewed (commented):
A few thoughts

rohitagg2020 commented:
LGTM

kumaritanushree commented:
LGTM

100mik commented on test/e2e/template_test.go:
```diff
@@ -276,6 +275,8 @@ func checkChangesOutput(t *testing.T, actualOutput, expectedOutput string) {
 	// (metadata.managedFields was added and threw off all lines numbers)
 	diffLinesRegexp := regexp.MustCompile(`(?m:^\s*(\d{1,3}\s*|\d{1,3},\s*\d{1,3}|\d{1,3}) [\-+ ])`)
 	actualOutput = diffLinesRegexp.ReplaceAllString(actualOutput, "-linesss-")
+	diffLinesRegexp = regexp.MustCompile(` *\d{1,3}, *\d{1,3}`)
```
What prompted us to make this change?

praveenrewar replied:
This was made to cover `-linesss-` from above.

100mik replied:
The empty line you mean?

praveenrewar replied:
The empty line with just the line number.

100mik replied:
Makes sense!

100mik commented on pkg/kapp/diff/change.go:
```diff
@@ -59,8 +63,11 @@ func NewChange(existingRes, newRes, appliedRes ctlres.Resource) *ChangeImpl {
 		appliedRes = appliedRes.DeepCopy()
 	}
+	if clusterOriginalRes != nil {
+		clusterOriginalRes = clusterOriginalRes.DeepCopy()
```
The motivation here is that the original resource should not be changed.
This might be unnecessary here as we do not really modify `ClusterOriginalResource` but that said it does not hurt and I am fine with having this as a safeguard to prevent future changes from breaking things.
(No action item here, just tracking thoughts - this comment can be resolved)

100mik reviewed (commented):
Nothing blocking here, just getting some clarity

100mik reviewed (approved):
LGTM!

sethiyash reviewed (approved):
LGTM!

kumaritanushree commented:
LGTM
